spec/app/ics-030-middleware/README.md

```diff
@@ -0,0 +1,237 @@
+      app.OnChanOpenTry(
+          order,
+          connectionHops,
+          portIdentifier,
+          channelIdentifier,
+          counterpartyPortIdentifier,
+          counterpartyChannelIdentifier,
+          cpAppVersion, // note we only pass counterparty app version here
+          appVersion, // only pass app version
+      )
+}
+
+function onChanOpenAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  version: string) {
+      middlewareVersion, appVersion = splitMiddlewareVersion(version)
+      if !isCompatible(middlewareVersion) {
+          return error
+      }
+      doCustomLogic()
+      
+      // call the underlying applications OnChanOpenTry callback
+      app.OnChanOpenAck(portIdentifier, channelIdentifier, appVersion)
+}
+
+function OnChanOpenConfirm(
+    portIdentifier: Identifier,
+    channelIdentifier: Identifier) {
+    doCustomLogic()
+
+    app.OnChanOpenConfirm(portIdentifier, channelIdentifier)
+}
+
+function splitMiddlewareVersion(version: string): []string {
+    splitVersions = split(version,  ":")
+    middlewareVersion = version[0]
+    appVersion = join(version[1:], ":")
+    return []string{middlewareVersion, appVersion}
+}
+```
+
+NOTE: Middleware that does not need to negotiate with a counterparty middleware on the remote stack will not implement the version splitting and negotiation, and will simply perform its own custom logic on the callbacks without relying on the counterparty behaving similarly.
+
+#### Packet Callbacks
+
+```typescript
+function onRecvPacket(packet: Packet, relayer: string): bytes {
+    doCustomLogic()
+
+    app_acknowledgement = app.onRecvPacket(packet, relayer)
+
+    // middleware may modify ack
+    ack = doCustomLogic(app_acknowledgement)
+   
+    return marshal(ack)
+}
+
+function onAcknowledgePacket(packet: Packet, acknowledgement: bytes, relayer: string) {
+    doCustomLogic()
+
+    // middleware may modify ack
+    app_ack = getAppAcknowledgement(acknowledgement)
+
+    app.OnAcknowledgePacket(packet, app_ack, relayer)
+
+    doCustomLogic()
+}
+
+function onTimeoutPacket(packet: Packet, relayer: string) {
+    doCustomLogic()
+
+    app.OnTimeoutPacket(packet, relayer)
+
+    doCustomLogic()
+}
+
+function onTimeoutPacketClose(packet: Packet, relayer: string) {
+    doCustomLogic()
+
+    app.onTimeoutPacketClose(packet, relayer)
+
+    doCustomLogic()
+}
+```
+
+NOTE: Middleware may do pre- and post-processing on underlying application data for all IBC Module callbacks defined in ICS-26.
+
+#### ICS-4 Wrappers
+
+```typescript
+function writeAcknowledgement(
+  packet: Packet,
+  acknowledgement: bytes) {
+    // middleware may modify acknowledgement
+    ack_bytes = doCustomLogic(acknowledgement)
+
+    return ics4.writeAcknowledgement(packet, ack_bytes)
+}
+```
+
+```typescript
+function sendPacket(app_packet: Packet) {
+    // middleware may modify packet
+    packet = doCustomLogic(app_packet)
+
+    return ics4.sendPacket(packet)
```

Comment:
If I am not wrong, we assume that chain of middleware wrappers function as a processing pipeline only upon message reception, but not in other direction (on packet and ack send). Wouldn't it be cleaner if we go through pipeline in both directions?

Reply:
This is explaining exactly that concept. On packet/ack receiving; processing goes from ICS-4 to top-layer middleware, through the chain, to base app. On packet/ack sending; processing goes from base application, from lowest to top middleware, and then to ics4.

So the pipeline exists in both directions. The above is an example of middleware that gets called by an underlying base application's `SendPacket` function. Note it is calling ics4, not the base app